### PR TITLE
simplify argument rector

### LINKED ISSUE #398: Let Rector\Rector\Dynamic\ArgumentRector add argument
Based on #353 

```yaml
    # @todo, use this and add new argument to a method
    Rector\Rector\Dynamic\ArgumentRector:
        $argumentByMethod:
             # Add state argument to the getLastPayment($state), where $state is target last payment state. getLastPayment(PaymentInterface::STATE_NEW)
            -
                type: 'add'
                class: 'Sylius\Component\Core\Model\OrderInterface'
                method: 'getLastPayment'
                position: 0 # as append | default value is 0, config argument need not be listed
                typehint: false # default value is false, config argument need not be listed
                property: '$state'
                default_value: PaymentInterface::STATE_NEW
```

### PR DESCRIPTION
**Pull request by mssimi**
Closes #398.

I think ArgumentRector had too complex configuration, I tried to simplify that. Now it has only two strategies  that depends on which argument is provided.

To add argument specify `position` and `replacement_value`.
To change argument specify `position` and `replacement_value`.
To remove argument specify `position` and `replacement_value: ~`.
To change with map specify `position` and `replacement_map`.

### CODE REVIEW

File: tests/Rector/Dynamic/ArgumentRector/config.yml [hunk 1]
```diff
@@ -5,30 +5,27 @@ services:
...
-                default_value: 0
+                replacement_value: 0
 
             # remove argument
             -
                 class: 'Doctrine\ORM\Persisters\Entity\AbstractEntityInheritancePersister'
                 method: 'getSelectJoinColumnSQL'
                 position: 4
-                type: 'removed'
+                replacement_value: ~
```
Comment (TomasVotruba) [inline tests/Rector/Dynamic/ArgumentRector/config.yml]: Could you also add test for `replacement_value: null`?

File: src/Configuration/Rector/ArgumentReplacerRecipeFactory.php [hunk 2]
```diff
@@ -55,4 +50,17 @@ private function ensureHasKey(array $data, string $key): void
...
     }
+
+    /**
+     * @param mixed[] $data
+     */
+    private function ensureSingleReplacementStrategy(array $data): void
+    {
+        if (array_key_exists('replacement_value', $data) && isset($data['replacement_map'])) {
+            throw new InvalidRectorConfigurationException(sprintf(
+                'Configuration for "%s" Rector should have single replacement strategy.',
```
Comment (TomasVotruba) [inline src/Configuration/Rector/ArgumentReplacerRecipeFactory.php]: What about printing those strategies as well?

**Comment by TomasVotruba** [comment]
Just a side thought: I wonder how to incorporate this to `RectorDefinition`, so user can read how to use.

File: tests/Rector/Dynamic/ArgumentRector/config.yml (hunk 1 above)
Comment (mssimi) [reply tests/Rector/Dynamic/ArgumentRector/config.yml]: Isn't tilde(~) equals to null in yml?

Comment (TomasVotruba) [reply tests/Rector/Dynamic/ArgumentRector/config.yml]: I don't know. How would `null` value be added then?

Comment (mssimi) [reply tests/Rector/Dynamic/ArgumentRector/config.yml]: Damn:D

File: src/Configuration/Rector/ArgumentReplacerRecipeFactory.php [hunk 3]
```diff
@@ -58,7 +58,7 @@ private function ensureSingleReplacementStrategy(array $data): void
     {
         if (array_key_exists('replacement_value', $data) && isset($data['replacement_map'])) {
             throw new InvalidRectorConfigurationException(sprintf(
-                'Configuration for "%s" Rector should have single replacement strategy.',
+                'Configuration for "%s" Rector should have single replacement strategy. Specify only one of "replacement_value" and "replacement_map"',
```
Comment (TomasVotruba) [inline src/Configuration/Rector/ArgumentReplacerRecipeFactory.php]: "and" seems like I should use both at the same time.

```diff
-Specify only one of "replacement_value" and "replacement_map"
+Specify only one of "replacement_value" or "replacement_map"
```

File: src/Configuration/Rector/ArgumentReplacerRecipeFactory.php [hunk 4]
```diff
@@ -58,7 +58,7 @@ private function ensureSingleReplacementStrategy(array $data): void
     {
         if (array_key_exists('replacement_value', $data) && isset($data['replacement_map'])) {
```
Comment (TomasVotruba) [inline src/Configuration/Rector/ArgumentReplacerRecipeFactory.php]: Btw, why 2 different functions for same thing? It's too hard to read


```php
isset($data['replacement_map'], $data['replacement_value'])
```

Comment (mssimi) [reply src/Configuration/Rector/ArgumentReplacerRecipeFactory.php]: isset returns false if key exist and value is null while array_key_exists returns true.

File: docs/AllRectorsOverview.md [hunk 5]
```diff
@@ -20,10 +20,10 @@ Turns parent EntityRepository class to constructor dependency
 
 ```diff
 namespace App\Repository;
- 
+
```
Comment (carusogabriel) [inline docs/AllRectorsOverview.md]: This PR can be rebased after #450, that also removed this spaces

File: src/Configuration/Rector/ArgumentReplacerRecipeFactory.php (hunk 4 above)
Comment (TomasVotruba) [reply src/Configuration/Rector/ArgumentReplacerRecipeFactory.php]: Oh, didn't know that.

Maybe group them into clear method `hasArraysKeys($array, $keys)`?

**Comment by mssimi** [comment]
After discussion on hangouts. This Rector is maybe too abstract, maybe we should split it to multiple Rectors. What Rectors do you propose?

**Comment by TomasVotruba** [comment]
I'd start with the one we have already cover:

- ArgumentAdderRector
- ArgumentValueReplacerRector
- ArgumentRemoverRector

**Comment by mssimi** [comment]
Is ArgumentValueReplacerRector meant to mimic replacement_map(aka TYPE_REPLACED_DEFAULT_VALUE)?

**Comment by TomasVotruba** [comment]
@mssimi Yes, exactly that

**Comment by mssimi** [comment]
Thx for tip, is it ok this way?

**Comment by TomasVotruba** [comment]
Also resolve this please: https://github.com/rectorphp/rector/issues/398#issuecomment-385353338

**Comment by mssimi** [comment]
I will check silius thing later today.

File: src/Rector/Dynamic/AbstractArgumentRector.php [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+     */
+    final public function setMethodCallAnalyzer(MethodCallAnalyzer $methodCallAnalyzer): void
+    {
+        $this->methodCallAnalyzer = $methodCallAnalyzer;
+    }
+
+    /**
+     * @required
+     */
+    final public function setClassMethodAnalyzer(ClassMethodAnalyzer $classMethodAnalyzer): void
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply src/Rector/Dynamic/AbstractArgumentRector.php]: I would not worry about it

**Comment by mssimi** [comment]
As of sylius should we extend argument rectors to support typehints or just chain it with `ParentTypehintedArgumentRector`?

**Comment by TomasVotruba** [comment]
> Just chain it with ParentTypehintedArgumentRector

This the style we use now. So this

I'll merge this PR, so there is new clean space for work.

**Comment by TomasVotruba** [comment]
Thanks for great work on this!

File: config/level/symfony/symfony32.yml [hunk 7]
```diff
@@ -1,10 +1,9 @@
...
-    Rector\Rector\Dynamic\ArgumentRector:
+    Rector\Rector\Dynamic\ArgumentAdderRector:
         $argumentChangesByMethodAndType:
             -
                 class: Symfony\Component\DependencyInjection\ContainerBuilder
                 method: addCompilerPass
                 position: 2
-                type: added
-                default_value: 0
+                replacement_value: 0
```
Comment (TomasVotruba) [inline config/level/symfony/symfony32.yml]: Maybe `default_value` is better here since there is no replacement?

File: config/level/symfony/symfony33.yml [hunk 8]
```diff
@@ -1,26 +1,23 @@
...
                 position: 2
-                default_value: 0
+                replacement: 0
             -
                 class: 'Symfony\Component\DependencyInjection\Compiler\ServiceReferenceGraph'
                 method: 'connect'
                 position: 6
-                type: 'changed'
-                default_value: false
+                replacement: false
```
Comment (TomasVotruba) [inline config/level/symfony/symfony33.yml]: Same on these 3 places
